File: Backend/coreservices/Inventory/src/main/java/mth/controller/AnalyticsController.java
package mth.controller;

import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import mth.dto.*;
import mth.service.AnalyticsService;
import mth.service.InventoryService;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/analytics")
public class AnalyticsController {

    private final AnalyticsService analyticsService;
    private final InventoryService inventoryService;

    public AnalyticsController(AnalyticsService analyticsService, InventoryService inventoryService) {
        this.analyticsService = analyticsService;
        this.inventoryService = inventoryService;
    }


    @GetMapping("/dashboard")
    public ResponseEntity<ApiResponseDTO<DashboardMetricsDTO>> getDashboardMetrics() {
        DashboardMetricsDTO metrics = analyticsService.getDashboardMetrics();
        return ResponseEntity.ok(ApiResponseDTO.success(metrics, "Dashboard metrics retrieved"));
    }

    @GetMapping("/categories")
    public ResponseEntity<ApiResponseDTO<List<CategoryAnalyticsDTO>>> getCategoryAnalytics() {
        List<CategoryAnalyticsDTO> analytics = analyticsService.getCategoryAnalytics();
        return ResponseEntity.ok(ApiResponseDTO.success(analytics, "Category analytics retrieved"));
    }

    @GetMapping("/stock-alerts")
    public ResponseEntity<ApiResponseDTO<List<StockAlertDTO>>> getStockAlerts() {
        List<StockAlertDTO> alerts = analyticsService.getStockAlerts();
        return ResponseEntity.ok(ApiResponseDTO.success(alerts, "Stock alerts retrieved"));
    }

    @GetMapping("/inventory-summary")
    public ResponseEntity<ApiResponseDTO<Map<String, Object>>> getInventorySummary() {
        Map<String, Object> summary = analyticsService.getInventorySummary();
        return ResponseEntity.ok(ApiResponseDTO.success(summary, "Inventory summary retrieved"));
    }

    @GetMapping("/low-stock")
    public ResponseEntity<ApiResponseDTO<List<InventoryItemResponseDTO>>> getLowStockItems() {
        List<InventoryItemResponseDTO> items = inventoryService.getLowStockItems();
        return ResponseEntity.ok(ApiResponseDTO.success(items, "Low stock items retrieved"));
    }

    @GetMapping("/sales-metrics")
    public ResponseEntity<ApiResponseDTO<List<SalesMetricsDTO>>> getSalesMetrics(@RequestParam(defaultValue = "7") int days) {
        List<SalesMetricsDTO> metrics = analyticsService.getSalesMetrics(days);
        return ResponseEntity.ok(ApiResponseDTO.success(metrics, "Sales metrics retrieved"));
    }

    @GetMapping("/top-selling-items")
    public ResponseEntity<ApiResponseDTO<List<Map<String, Object>>>> getTopSellingItems(@RequestParam(defaultValue = "5") int limit) {
        List<Map<String, Object>> items = analyticsService.getTopSellingItems(limit);
        return ResponseEntity.ok(ApiResponseDTO.success(items, "Top selling items retrieved"));
    }

    @GetMapping("/category/{categoryName}")
    public ResponseEntity<ApiResponseDTO<Map<String, Object>>> getInventoryByCategory(@PathVariable String categoryName) {
        List<InventoryItemResponseDTO> items = inventoryService.getItemsByCategory(categoryName);
        Map<String, Object> result = new HashMap<>();
        result.put("category", categoryName);
        result.put("items", items);
        result.put("total", items.size());
        return ResponseEntity.ok(ApiResponseDTO.success(result, "Category items retrieved"));
    }

    @GetMapping("/turnover-metrics")
    public ResponseEntity<ApiResponseDTO<Map<String, Object>>> getTurnoverMetrics() {
        Map<String, Object> metrics = analyticsService.calculateTurnoverMetrics();
        return ResponseEntity.ok(ApiResponseDTO.success(metrics, "Turnover metrics retrieved"));
    }
}

File: Backend/gateway/Controllers/analyticsController.py
from fastapi import APIRouter, Header, Query
import httpx

router = APIRouter(prefix="/api/analytics")

SPRING_URL = "http://localhost:8083/"


@router.get("/dashboard")
async def get_dashboard_metrics(Authorization: str = Header(...)):
    async with httpx.AsyncClient() as client:
        response = await client.get(
            SPRING_URL + "api/analytics/dashboard",
            headers={"Authorization": Authorization}
        )

    return response.json()


@router.get("/categories")
async def get_category_analytics(Authorization: str = Header(...)):
    async with httpx.AsyncClient() as client:
        response = await client.get(
            SPRING_URL + "api/analytics/categories",
            headers={"Authorization": Authorization}
        )

    return response.json()


@router.get("/stock-alerts")
async def get_stock_alerts(Authorization: str = Header(...)):
    async with httpx.AsyncClient() as client:
        response = await client.get(
            SPRING_URL + "api/analytics/stock-alerts",
            headers={"Authorization": Authorization}
        )

    return response.json()


@router.get("/inventory-summary")
async def get_inventory_summary(Authorization: str = Header(...)):
    async with httpx.AsyncClient() as client:
        response = await client.get(
            SPRING_URL + "api/analytics/inventory-summary",
            headers={"Authorization": Authorization}
        )

    return response.json()


@router.get("/low-stock")
async def get_low_stock_items(Authorization: str = Header(...)):
    async with httpx.AsyncClient() as client:
        response = await client.get(
            SPRING_URL + "api/analytics/low-stock",
            headers={"Authorization": Authorization}
        )

    return response.json()


@router.get("/sales-metrics")
async def get_sales_metrics(days: int = Query(7), Authorization: str = Header(...)):
    async with httpx.AsyncClient() as client:
        response = await client.get(
            SPRING_URL + "api/analytics/sales-metrics",
            params={"days": days},
            headers={"Authorization": Authorization}
        )

    return response.json()


@router.get("/top-selling-items")
async def get_top_selling_items(limit: int = Query(5), Authorization: str = Header(...)):
    async with httpx.AsyncClient() as client:
        response = await client.get(
            SPRING_URL + "api/analytics/top-selling-items",
            params={"limit": limit},
            headers={"Authorization": Authorization}
        )

    return response.json()


@router.get("/category/{category_name}")
async def get_inventory_by_category(category_name: str, Authorization: str = Header(...)):
    async with httpx.AsyncClient() as client:
        response = await client.get(
            f"{SPRING_URL}api/analytics/category/{category_name}",
            headers={"Authorization": Authorization}
        )

    return response.json()


@router.get("/turnover-metrics")
async def get_turnover_metrics(Authorization: str = Header(...)):
    async with httpx.AsyncClient() as client:
        response = await client.get(
            SPRING_URL + "api/analytics/turnover-metrics",
            headers={"Authorization": Authorization}
        )

    return response.json()
